File: packages/uc04/src/uc04/core/fingerprint.py
"""Explanation fingerprinting - the mechanism behind "a paraphrase is a repeat".

Two checks, both applied:

1. an exact hash of the canonicalised content-token stream, so reformatting or reordering an
   explanation cannot present it as new;
2. Jaccard similarity against every earlier attempt for the same concept in the same session,
   so a reworded but substantively identical explanation is rejected too.

The fingerprint is content-only: headings and framing labels are stripped by the tokeniser, so
"same explanation under a different heading" collides exactly as it should.
"""

from __future__ import annotations

from collections.abc import Iterable
from dataclasses import dataclass

from ..domain.models import FramingAttempt
from .text import content_tokens, jaccard, stable_hash
from .thresholds import PARAPHRASE_SIMILARITY_THRESHOLD


@dataclass(frozen=True)
class Fingerprint:
    value: str
    tokens: tuple[str, ...]


@dataclass(frozen=True)
class RepeatVerdict:
    is_repeat: bool
    #: "exact" when the hashes collide, "paraphrase" when similarity crossed the threshold.
    reason: str | None
    similarity: float


def fingerprint(text: str) -> Fingerprint:
    canonical = tuple(sorted(set(content_tokens(text))))
    return Fingerprint(value=stable_hash(" ".join(canonical)), tokens=canonical)


def is_repeat(
    candidate: Fingerprint,
    previous: Iterable[FramingAttempt],
    threshold: float = PARAPHRASE_SIMILARITY_THRESHOLD,
) -> RepeatVerdict:
    highest = 0.0
    for attempt in previous:
        if attempt.fingerprint == candidate.value:
            return RepeatVerdict(is_repeat=True, reason="exact", similarity=1.0)
        similarity = jaccard(candidate.tokens, attempt.fingerprint_tokens)
        highest = max(highest, similarity)
    if highest >= threshold:
        return RepeatVerdict(is_repeat=True, reason="paraphrase", similarity=round(highest, 4))
    return RepeatVerdict(is_repeat=False, reason=None, similarity=round(highest, 4))

File: reference/uc04-typescript/src/core/explanation/fingerprint.ts
import { contentTokens, jaccard, stableHash } from '../text';

/**
 * Stable explanation fingerprinting.
 *
 * Two mechanisms, both used:
 *  1. `fingerprint`  - an exact, stable hash of the normalized content-token stream. Identical
 *     (or trivially reformatted) explanations collide here.
 *  2. `tokens`       - the normalized token set, used for NEAR-duplicate detection via Jaccard
 *     similarity, so a reworded but substantively identical explanation is still rejected.
 *
 * The fingerprint is content-only: it deliberately ignores framing labels and headings so that
 * "same explanation with a different heading" cannot slip through as new.
 */
export interface ExplanationFingerprint {
  fingerprint: string;
  tokens: string[];
}

/** Similarity at or above this is treated as "effectively identical". */
export const NEAR_DUPLICATE_THRESHOLD = 0.82;

export function fingerprintExplanation(text: string): ExplanationFingerprint {
  const tokens = contentTokens(text);
  // Sorted + de-duplicated so word order changes alone do not create a "new" explanation.
  const canonical = Array.from(new Set(tokens)).sort();
  return {
    fingerprint: stableHash(canonical.join(' ')),
    tokens: canonical,
  };
}

export interface DuplicateVerdict {
  isDuplicate: boolean;
  /** 'EXACT' when fingerprints collide, 'NEAR' when similarity crosses the threshold. */
  reason: 'EXACT' | 'NEAR' | null;
  similarity: number;
}

export function isEffectivelyIdentical(
  candidate: ExplanationFingerprint,
  previous: readonly { explanation_fingerprint: string; fingerprint_tokens: string[] }[],
  threshold: number = NEAR_DUPLICATE_THRESHOLD,
): DuplicateVerdict {
  let maxSimilarity = 0;
  for (const attempt of previous) {
    if (attempt.explanation_fingerprint === candidate.fingerprint) {
      return { isDuplicate: true, reason: 'EXACT', similarity: 1 };
    }
    const similarity = jaccard(candidate.tokens, attempt.fingerprint_tokens);
    if (similarity > maxSimilarity) maxSimilarity = similarity;
  }
  if (maxSimilarity >= threshold) {
    return { isDuplicate: true, reason: 'NEAR', similarity: round(maxSimilarity) };
  }
  return { isDuplicate: false, reason: null, similarity: round(maxSimilarity) };
}

function round(value: number): number {
  return Math.round(value * 10000) / 10000;
}
